compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs

```diff
@@ -1337,16 +1342,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         }
                     }
                     GenericParamDefKind::Const { has_default } => {
-                        if !infer_args
-                            && has_default
-                            && !tcx.has_attr(param.def_id, sym::rustc_host)
-                        {
-                            tcx.const_param_default(param.def_id)
-                                .instantiate(tcx, args.unwrap())
-                                .into()
-                        } else {
-                            self.fcx.var_for_def(self.span, param)
+                        if has_default {
+                            if tcx.has_attr(param.def_id, sym::rustc_host) {
+                                return self.fcx.var_for_effect(param);
+                            } else if !infer_args {
+                                return tcx
+                                    .const_param_default(param.def_id)
+                                    .instantiate(tcx, args.unwrap())
+                                    .into();
+                            }
                         }
```

Comment:
Pre-existing: Can you comment why this logic is this way. `has_default && !infer_args`, but we're *still* instantiating a var for `#[rustc_host]` const params?

Reply:
yep